server/src/main/java/org/opensearch/cluster/NodeConnectionsService.java

```diff
@@ -174,6 +175,17 @@ public void disconnectFromNodesExcept(DiscoveryNodes discoveryNodes) {
             for (final DiscoveryNode discoveryNode : nodesToDisconnect) {
                 runnables.add(targetsByNode.get(discoveryNode).disconnect());
             }
+
+            // There might be some stale nodes that are in pendingDisconnect set from before but are not connected anymore
+            // This code block clears the pending disconnect for these nodes to avoid permanently blocking node joins
+            // This situation should ideally not happen
+            transportService.removePendingDisconnections(
```

Comment:
Right, this is correct.
This check was meant to be specifically to remove nodes that have pending disconnect marked but don't have entries present in `targetsByNode` list, so disconnect would not ever be called for these nodes.

Will update this to also have a check ensuring the node is not there in targetsByNode.

`.filter(discoveryNode -> !discoveryNodes.nodeExists(discoveryNode) && !targetsByNode.containsKey(discoveryNode))`

Reply:
Why would it happen that the node is disconnected but it still part of pendingDisconnect list ?

Reply:
This is the scenario we were seeing in `CoordinatorTests` failures

This was due to the custom extension of ClusterApplierService in these tests which are completely skipping connectToNodes logic:

```
@Override
        protected void connectToNodesAndWait(ClusterState newClusterState) {
            // don't do anything, and don't block
        }
```

In this scenario, `targetsByNode` never gets reconstructed when a new node is connected.

In state commit flow, after `connectToNodesAndWait`  is called, `nodeConnectionsService.disconnectFromNodesExcept(newClusterState.nodes());` is then called 

This function uses `targetsByNode` to decide which nodes to disconnect. So in this case, we were seeing that during node-left task for a node, the pendingDisconnections list was never getting cleared up because it didn't have an entry in `targetsByNode`